### ISSUE TITLE
failed to override main function on Ubuntu 22.04

### ISSUE THREAD

**Issue by 1saac-k** [Snaipe/Criterion#495]
Hi.
I installed criterion via apt on Ubuntu 22.04. (libcriterion-dev, version: `2.4.0-rc-18-g1cc3911-1`)
In order to support custom arguments, I added the main function according to the below manual.

> https://criterion.readthedocs.io/en/master/internal.html

Then the following error appeared.

``` bash
labuser@openstack1:~/criterion-test$ cat main.c
#include <criterion/criterion.h>

Test(foo, bar) { }

int main(int argc, char *argv[]) {
    struct criterion_test_set *tests = criterion_initialize();

    int result = 0;
    if (criterion_handle_args(argc, argv, true))
        result = !criterion_run_all_tests(tests);

    criterion_finalize(tests);
    return result;
}
labuser@openstack1:~/criterion-test$ gcc -lcriterion -o main main.c
labuser@openstack1:~/criterion-test$ ./main
criterion: Re-entering criterion from a test worker. This is a catastrophic bug, please report it on the issue tracker.
Bailing out to avoid fork-bombing the system.
criterion: Could not spawn test instance: Protocol error
Aborted (core dumped)
```

As you probably remember, the error message has been added in the issue below.
- https://github.com/Snaipe/Criterion/issues/247
  - https://github.com/Snaipe/Criterion/commit/57a518fac021aa298bb6b3f91beb914d24bdf1c2
  - https://github.com/Snaipe/BoxFort/commit/926bd4ce968592dbbba97ec1bb9aeca3edf29b0d

If my understanding is correct, in normal operation, boxfort manipulates the main function I added, and when the main function is executed again through fork & execve, the bxfi_exec function should be called by the manipulation.

However, it seems that boxfort manipulated the wrong main function (the main function inside the criterion), and due to this, when the main function is executed, the criterion_run_all_tests function is executed again and an error message is output as an exception handling to prevent falling into an infinite loop.

I'm not familiar with the internals of criterion and boxfort, so you can disregard my assumptions.

Strangely, there is no problem with the package downloaded from github.com (here) and the package I built myself.
To experiment with different packages, I used LD_LIBRARY_PATH as in the example below.

``` bash
# using apt package -> ERROR
LD_LIBRARY_PATH=/usr/lib/x86_64-linux-gnu ./main

# using compiled package (g1cc3911 version) -> OK
LD_LIBRARY_PATH=/usr/local/lib/x86_64-linux-gnu ./main

# using downloaded package -> all OK
LD_LIBRARY_PATH=$HOME/ws/criterion-2.4.0-rc-18-g1cc3911/lib ./main
LD_LIBRARY_PATH=$HOME/ws/criterion-2.4.0-rc-12-g4cccb18/lib ./main
LD_LIBRARY_PATH=$HOME/ws/criterion-2.4.0/lib ./main
LD_LIBRARY_PATH=$HOME/ws/criterion-2.4.1/lib ./main
```

So I thought it was not a version issue.
I tried compiling using clang instead of gcc as shown below, but there is still no problem with the package I built myself.
``` bash
meson setup --wipe build
env CC=clang meson build
ninja -C build
sudo ninja -C build install
```

I wonder if you can reproduce it too. I've experimented with multiple servers.
Thank you for your hard work!

**Comment by MrAnno** [comment]
Hi @kyet,

Thanks for reporting this issue.
I've reproduced the same using the official Ubuntu package.

So far I don't see what causes it, perhaps a hardening compiler flag.

**Comment by MrAnno** [comment]
Note: On Debian "testing", it works as expected.

**Comment by MrAnno** [comment]
Yep, this is caused by the `-Wl,-Bsymbolic-functions` linker flag that is used in Ubuntu builds.

**Comment by MrAnno** [comment]
I've submitted a bug report:
https://bugs.launchpad.net/ubuntu/+source/criterion/+bug/2018156
